### ISSUE TITLE
False positive `Optional` incompatible check

### ISSUE THREAD

**Issue by scruel** [microsoft/pyright#8092]
**Describe the bug**
The `__init__` function has an optional parameter and set as a property for the object, however in the later implementation, this value won't be `None`, but pyright will still consider this property as optional, and gives the following error:
```
Argument of type "str | None" cannot be assigned to parameter "text" of type "str" in function "process"
  Type "str | None" is incompatible with type "str"
    "None" is incompatible with "str"Pylance
```

**Code or Screenshots**

```python
from typing import Optional

class A():
    def __init__(self, text: Optional[str] = None,
                 value: Optional[int] = None):
        if text is None and value is None:
            raise ValueError()
        self._text = text
        self._value = value

    def process(self, text: str):
        pass

    def get_value(self):
        if self._value is None:
            self._value = self.process(self._text)
        return self._value

```

**VS Code extension or command-line**
```
Version: 1.90.0 (user setup)
Commit: 89de5a8d4d6205e5b11647eb6a74844ca23d2573
Date: 2024-06-04T19:33:54.889Z
Electron: 29.4.0
ElectronBuildId: 9593362
Chromium: 122.0.6261.156
Node.js: 20.9.0
V8: 12.2.281.27-electron.0
OS: Windows_NT x64 10.0.22631
```

**Comment by erictraut** [comment]
Pyright is working correctly here, so I don't consider this a bug.

The `process` method indicates that it requires a `str` argument, but you are passing it a value from `self._text`. The instance variable `_text` has no type declaration, so its type is inferred from all assignments to it within the class. There is only one such assignment, where it is assigned a value with type `str | None`.

The type for instance variable `_text` is not conditioned on the type of instance variable `_value`. That's not how static type checking works. These are two independent variables with independent types. Both of them have have the inferred type `str | None`.

For more information about how static typing works, refer to pyright's [documentation](https://microsoft.github.io/pyright/#/type-concepts).

You can avoid the type violation by adding an assert to the `get_value` method. This will document your assumption (to both the type checker and to human readers) that if `self._value` is `None`, then `self._text` cannot be `None`.
```python
    def get_value(self):
        if self._value is None:
            assert self._text is not None
            self._value = self.process(self._text)
        return self._value
```

**Comment by scruel** [comment]
`assert self._text is not None` is redundant and will only be used for passing pyright check it user won't try to execute `self._text=self._text=None`, though I understand why this can't be fixed now, and for robustness, I think I will impl this in another way, thanks though.
